parlai/agents/transformer/image_polyencoder.py

```diff
@@ -0,0 +1,215 @@
+#!/usr/bin/env python3
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+# hack to make sure -m transformer/generator works as expected
+"""
+Poly-encoder agent that ingests image features.
+"""
+
+from typing import Any, Dict
+
+import torch
+
+from parlai.agents.image_seq2seq.modules import ContextWithImageEncoder
+from parlai.agents.transformer.modules import get_n_positions_from_options
+from parlai.agents.transformer.polyencoder import PolyencoderAgent, PolyEncoderModule
+from parlai.core.torch_agent import Batch
+from parlai.core.torch_image_agent import TorchImageAgent
+
+
+class ImagePolyencoderAgent(PolyencoderAgent, TorchImageAgent):
+    """
+    Poly-encoder Agent that ingests image features.
+
+    Agent that allows encoding image features and adding or concatenating them to the
+    context encoding.
+    """
+
+    @classmethod
+    def add_cmdline_args(cls, argparser):
+        """
+        Add command-line arguments specifically for this agent.
+        """
+        PolyencoderAgent.add_cmdline_args(argparser)
+        TorchImageAgent.add_cmdline_args(argparser)
+        agent = argparser.add_argument_group('Image Encoder Args')
+        agent.add_argument(
+            '--image-combination-mode',
+            type=str,
+            default='prepend',
+            choices=['add', 'append', 'prepend'],
+            help='How to combine image embedding (if used) with context embedding',
+        )
+        agent.add_argument(
+            '--n-image-tokens',
+            type=int,
+            default=1,
+            help=(
+                'Number of tokens that the image encoding will consist of (when adding '
+                'or prepending)'
+            ),
+        )
+        agent.set_defaults(reduction_type=None)
+        # This agent doesn't support any encoder output reductions
+        return agent
+
+    def build_model(self, states=None):
+        """
+        Return built model.
+        """
+        return ImagePolyencoderModule(self.opt, self.dict, self.NULL_IDX)
+
+    def batchify_image_features(self, batch: Batch) -> Batch:
```

Comment:
could you maybe elaborate a bit about how this differs from the `batchify_image_features` in `ImageSeq2seq`? if it's not _too_ much different, it might be good to include in the `TorchImageAgent`

Reply:
Okay, I've added a note about this. Yes, I agree that it'd be good to test out whether `ImageSeq2seq` really requires a separate batchify method now. I've created an issue for this: https://github.com/facebookresearch/ParlAI/issues/2461